lib/thread.c

```diff
@@ -41,6 +41,14 @@ DEFINE_MTYPE_STATIC(LIB, THREAD_STATS,  "Thread stats")
 #include <mach/mach_time.h>
 #endif
 
+static unsigned char wakebyte = 0x01;
+
+#define ACQUIRE_MASTER(m) \
+  do { \
+    while (pthread_mutex_trylock (&m->mtx)) \
+      write (m->io_pipe[1], &wakebyte, 1); \
+  } while (0);
+
```

Comment:
>This will block in write() after filling up the kernel's pipe buffer, i.e 4096 useless syscalls later.

No. The pipe should be set to nonblocking. My understanding is that when the buffer fills up it'll just return EAGAIN or similar and continue to spinlock. Correct me if I'm wrong. Clearly I missed one fcntl() call on the write fd; thank you for catching that.

>the design is seriously bad

🌮 💩 

>The better design is for `thread_fetch()` to not hold `m->mtx` during the `fd_select` call.

I went this direction initially.

>There is no reason for it to do so; none of the working data of the `fd_select` call can be modified by another thread.

This is incorrect and is precisely the problem I ran into. The thread master's fdset / pollfds are passed directly -- not copied -- into poll or select. For example:

```
  num = poll (m->handler.pfds, m->handler.pfdcount + m->handler.pfdcountsnmp,
              timeout);
```

thread_add_read_write() and thread_cancel() modify those. What's your proposed solution?

>If `thread_fetch` is restructured this way, the pipe-poke can be done (unconditionally) at mutex _release_ time instead. The purpose slightly changes: instead of poking to get the mutex, it's a poke to update the `fd_select` arguments to in-/exclude new items.

Sounds great. That is, of course, one of the dual purposes of this design - not only do we need to get around holding the lock during select / poll, we also need to restart that system call after we add new fd's to the set.

So, provided I put in that missing `fcntl()` on the write fd -- what are the remaining issues?